File: 009_Jump Game.cpp
/*
Jump Game

Given an array of non-negative integers nums, where nums[i] represents the maximum jump length from that position, determine if you can reach the last index starting from the first index.

Strategy: Greedy keep track of farthest reachable index.
*/

#include <vector>
#include <vector>
#include <iostream>
using namespace std;

bool canJump(const vector<int>& nums) {
    int reachable = 0;
    for (int i = 0; i < (int)nums.size(); ++i) {
        if (i > reachable) return false;
        reachable = max(reachable, i + nums[i]);
        if (reachable >= (int)nums.size() - 1) return true;
    }
    return false;
}


int main() {
    cout << boolalpha;
    cout << "CanJump 1: " << canJump({2,3,1,1,4}) << '\n';
    cout << "CanJump 2: " << canJump({3,2,1,0,4}) << '\n';
    cout << "CanJump 3: " << canJump({0}) << '\n';
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(1)
*/

File: 009_Jump Game.py
"""
Jump Game

Given an array of non-negative integers nums, where nums[i] is the max jump length from that position, determine if you can reach the last index.

Strategy: Greedy — maintain farthest reachable index.
"""

from typing import List

def can_jump(nums: List[int]) -> bool:
    reachable = 0
    for i, step in enumerate(nums):
        if i > reachable:
            return False
        reachable = max(reachable, i + step)
        if reachable >= len(nums) - 1:
            return True
    return True


if __name__ == '__main__':
    print('CanJump 1:', can_jump([2,3,1,1,4]))
    print('CanJump 2:', can_jump([3,2,1,0,4]))
    print('CanJump 3:', can_jump([0]))

# Time Complexity: O(n)
# Space Complexity: O(1)
